### PR TITLE
op-proposer, op-batcher: Support private keys in addition to mnemonics

### PR DESCRIPTION
**Pull request by mslipper**

### CODE REVIEW

File: op-proposer/flags/flags.go [hunk 1]
```diff
@@ -80,6 +80,12 @@ var (
 		Required: true,
 		EnvVar:   prefixEnvVar("L2_OUTPUT_HD_PATH"),
 	}
+	PrivateKey = cli.StringFlag{
+		Name:     "private-key",
+		Usage:    "The private key to use with the l2output wallet. Must notbe used with mnemonic.",
```
Comment (trianglesphere) [inline op-proposer/flags/flags.go:85]:
```suggestion
		Usage:    "The private key to use with the l2output wallet. Must not be used with mnemonic.",
```

**Review by trianglesphere — APPROVED** [review]

File: op-batcher/batch_submitter.go [hunk 2]
```diff
@@ -142,19 +144,31 @@ func NewBatchSubmitter(cfg Config, l log.Logger) (*BatchSubmitter, error) {
...
-	if err != nil {
-		return nil, err
-	}
+	var sequencerPrivKey *ecdsa.PrivateKey
+	var addr common.Address
 
-	sequencerPrivKey, err := wallet.PrivateKey(acc)
-	if err != nil {
-		return nil, err
+	if cfg.PrivateKey == "" {
```
[the remark this answers is not in the document]
Comment (mslipper) [reply op-batcher/batch_submitter.go:155]: ah yeah nice catch, fixed

**Comment by tynes** [comment]
I think @protolambda wanted to move to using encrypted keystores
https://github.com/ethereum/go-ethereum/blob/a0b88ce869d6be243c6b1900d69b9401f2d2c092/cmd/ethkey/generate.go#L109
https://github.com/ethereum/go-ethereum/blob/a0b88ce869d6be243c6b1900d69b9401f2d2c092/accounts/keystore/key.go#L66

I think this is fine because it won't require any changes in tooling or the way we handle secrets

File: op-proposer/l2_output_submitter.go [hunk 3]
```diff
@@ -121,20 +126,33 @@ func NewL2OutputSubmitter(
...
 
 	ctx := context.Background()
+	var l2OutputPrivKey *ecdsa.PrivateKey
+	var err error
 
-	// Parse l2output wallet private key and L2OO contract address.
-	wallet, err := hdwallet.NewFromMnemonic(cfg.Mnemonic)
-	if err != nil {
-		return nil, err
+	if cfg.PrivateKey != "" && cfg.Mnemonic != "" {
```
Comment (tynes) [inline op-proposer/l2_output_submitter.go:132]: Implementation in the batch-submitter also checks that one of them is enabled
https://github.com/ethereum-optimism/optimism/blob/0e7184b0f09eee193608e78cd0feb1431a2b145a/batch-submitter/config.go#L259

Comment (mslipper) [reply op-proposer/l2_output_submitter.go:132]: Do you want me to change the implementation here to match the one in the BSS?

Comment (tynes) [reply op-proposer/l2_output_submitter.go:132]: Not super opinionated but could reduce problems at deploy time

**Comment by protolambda** [comment]
Maybe ideally we define one go module that can support the loading of private keys, in various ways:
- mnemonic + path
- private key
- encrypted keystore + password

That way we don't duplicate security-critical code, make it easy to audit, and easy to use.

Also ideally with discouraged use of mnemonic and private key. And warnings if you put the mnemonic, private key, or password in CLI instead of env var. (Since shell history / log streaming may log the CLI command and leak it)

I've been thinking we should have an `op-service` go module, for exactly this type of thing. As well as logging configuration, metrics, and pprof.

I'm fine with merging this PR as-is, but it would be nice if we can refactor it all into a common `op-service` module later.

File: op-proposer/l2_output_submitter.go (hunk 3 above)
Comment (mslipper) [reply op-proposer/l2_output_submitter.go:132]: Will leave as-is for now.

**Comment by mslipper** [comment]
I agree with the `op-service` module. We have a lot of duplication among our various Go projects and it would be great to clean it up. We'll also need a separate "Signer" function to support things like Cloud HSMs in prod.

I'll merge this as-is, but will ticket the `op-service` work so we can get started on it.

**Pull request merged**